File: ios/Classes/MusicFeatureAnalyzerPlugin.swift
//
// MusicFeatureAnalyzerPlugin.swift
//
// Flutter plugin for native audio metadata on iOS (AVFoundation).
// Registration: Flutter discovers this plugin via pubspec.yaml (flutter.plugin.platforms.ios.pluginClass)
// and calls register(with:) automatically — no manual AppDelegate changes needed.
//

import Flutter
import UIKit
import AVFoundation

public class MusicFeatureAnalyzerPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?

    /// Called by Flutter's generated plugin registrant when the app starts.
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.music_feature_analyzer/audio_metadata",
            binaryMessenger: registrar.messenger()
        )
        let instance = MusicFeatureAnalyzerPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "verifyConnection" {
            let response: [String: Any] = [
                "connected": true,
                "channelName": "com.music_feature_analyzer/audio_metadata",
                "loadingMode": "IOS_PLATFORM",
                "handlerSet": channel != nil,
                "platform": "iOS",
                "pluginVersion": "1.0.3"
            ]
            result(response)
            return
        }

        guard let arguments = call.arguments as? [String: Any],
              let path = arguments["path"] as? String,
              !path.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            result(FlutterError(code: "INVALID_ARGUMENT",
                                message: "File path is required",
                                details: nil))
            return
        }

        switch call.method {
        case "getMetadata":
            getMetadata(filePath: path, result: result)
        case "getAlbumArt":
            getAlbumArt(filePath: path, result: result)
        case "getAlbumArtMimeType":
            getAlbumArtMimeType(filePath: path, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func getMetadata(filePath: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async {
            var metadata: [String: Any?] = [:]

            do {
                let fileManager = FileManager.default
                guard fileManager.fileExists(atPath: filePath) else {
                    metadata["error"] = "File does not exist: \(filePath)"
                    metadata["mimeType"] = self.getMimeTypeFromExtension(filePath: filePath)
                    DispatchQueue.main.async { result(metadata) }
                    return
                }

                do {
                    if let fileSize = try fileManager.attributesOfItem(atPath: filePath)[.size] as? UInt64 {
                        metadata["fileSize"] = Int(min(fileSize, UInt64(Int32.max)))
                    }
                } catch {
                    metadata["fileSize"] = nil
                }

                let fileURL = URL(fileURLWithPath: filePath)
                let asset = AVAsset(url: fileURL)

                let duration = asset.duration
                let durationInSeconds = CMTimeGetSeconds(duration)
                if durationInSeconds.isFinite && !durationInSeconds.isNaN && durationInSeconds >= 0 {
                    metadata["duration"] = Int64(durationInSeconds * 1000)
                } else {
                    metadata["duration"] = nil
                }

                let common = self.extractCommonMetadata(from: asset)
                metadata.merge(self.sanitizeMetadata(common)) { _, new in new }

                // estimatedDataRate is in bits/sec; convert to kbps for Dart contract
                if let track = asset.tracks(withMediaType: .audio).first {
                    let bps = track.estimatedDataRate
                    metadata["bitrate"] = bps > 0 ? Int(bps / 1000) : nil
                }

                metadata["mimeType"] = self.getMimeTypeFromExtension(filePath: filePath)
                metadata["hasAlbumArt"] = self.hasAlbumArt(in: asset)

                DispatchQueue.main.async { result(metadata) }
            } catch {
                metadata["error"] = "Exception in getMetadata: \(error.localizedDescription)"
                if metadata["mimeType"] == nil {
                    metadata["mimeType"] = self.getMimeTypeFromExtension(filePath: filePath)
                }
                DispatchQueue.main.async { result(metadata) }
            }
        }
    }

    private func extractCommonMetadata(from asset: AVAsset) -> [String: Any?] {
        var metadata: [String: Any?] = [:]
        let commonMetadata = asset.commonMetadata

        for item in commonMetadata {
            guard let key = item.commonKey?.rawValue else { continue }
            let value = item.stringValue

            switch key {
            case AVMetadataKey.commonKeyTitle.rawValue:
                metadata["title"] = value
            case AVMetadataKey.commonKeyArtist.rawValue:
                metadata["artist"] = value
            case AVMetadataKey.commonKeyAlbumName.rawValue:
                metadata["album"] = value
            case "TPE2", AVMetadataKey.iTunesMetadataKeyAlbumArtist.rawValue:
                metadata["albumArtist"] = value
            case AVMetadataKey.commonKeyType.rawValue:
                if metadata["genre"] == nil { metadata["genre"] = value }
            case AVMetadataKey.id3MetadataKeyContentType.rawValue,
                 AVMetadataKey.iTunesMetadataKeyUserGenre.rawValue:
                metadata["genre"] = value
            case AVMetadataKey.commonKeyCreationDate.rawValue,
                 AVMetadataKey.id3MetadataKeyYear.rawValue:
                if let yearString = value {
                    metadata["year"] = extractYearFromString(yearString)
                    metadata["date"] = yearString
                }
            case AVMetadataKey.commonKeyCreator.rawValue,
                 AVMetadataKey.id3MetadataKeyComposer.rawValue:
                metadata["composer"] = value
            case "TEXT":
                metadata["writer"] = value
            case AVMetadataKey.id3MetadataKeyTrackNumber.rawValue:
                if let trackInfo = value { metadata["trackNumber"] = extractTrackNumber(trackInfo) }
            case "TPOS", AVMetadataKey.iTunesMetadataKeyDiscNumber.rawValue:
                if let discInfo = value { metadata["discNumber"] = extractDiscNumber(discInfo) }
            default:
                break
            }
        }

        for format in asset.availableMetadataFormats {
            let items = asset.metadata(forFormat: format)
            for item in items {
                if item.key as? String == AVMetadataKey.iTunesMetadataKeyUserGenre.rawValue,
                   metadata["genre"] == nil {
                    metadata["genre"] = item.stringValue
                }
            }
        }

        return metadata
    }

    /// Sanitize string values: treat "null" or empty as nil (matches Android behavior).
    private func sanitizeMetadata(_ metadata: [String: Any?]) -> [String: Any?] {
        var out: [String: Any?] = [:]
        for (key, value) in metadata {
            if let s = value as? String, (s.isEmpty || s == "null") {
                out[key] = nil
            } else {
                out[key] = value
            }
        }
        return out
    }

    private func extractYearFromString(_ dateString: String) -> String? {
        if let range = dateString.range(of: "\\d{4}", options: .regularExpression) {
            return String(dateString[range])
        }
        let parts = dateString.split(separator: "-")
        if let first = parts.first, first.count == 4 { return String(first) }
        let partsSlash = dateString.split(separator: "/")
        if let last = partsSlash.last, last.count == 4 { return String(last) }
        return dateString.isEmpty ? nil : dateString
    }

    private func extractTrackNumber(_ trackInfo: String) -> String? {
        let parts = trackInfo.split(separator: "/")
        guard let first = parts.first else { return nil }
        return String(first).isEmpty ? nil : String(first)
    }

    private func extractDiscNumber(_ discInfo: String) -> String? {
        let parts = discInfo.split(separator: "/")
        guard let first = parts.first else { return nil }
        return String(first).isEmpty ? nil : String(first)
    }

    private func getMimeTypeFromExtension(filePath: String) -> String {
        let ext = (filePath as NSString).pathExtension.lowercased()
        switch ext {
        case "mp3": return "audio/mpeg"
        case "m4a", "m4p", "m4b": return "audio/mp4"
        case "aac": return "audio/aac"
        case "wav": return "audio/wav"
        case "flac": return "audio/flac"
        case "ogg": return "audio/ogg"
        case "opus": return "audio/opus"
        case "aiff", "aif": return "audio/aiff"
        case "caf": return "audio/x-caf"
        case "3gp", "3gpp": return "audio/3gpp"
        case "amr": return "audio/amr"
        case "au": return "audio/basic"
        case "wma": return "audio/x-ms-wma"
        default: return "audio/mpeg"
        }
    }

    private func hasAlbumArt(in asset: AVAsset) -> Bool {
        let items = AVMetadataItem.metadataItems(
            from: asset.commonMetadata,
            filteredByIdentifier: .commonIdentifierArtwork
        )
        return !items.isEmpty
    }

    private func getAlbumArt(filePath: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async {
            let fileManager = FileManager.default
            guard fileManager.fileExists(atPath: filePath) else {
                DispatchQueue.main.async { result(nil) }
                return
            }

            let fileURL = URL(fileURLWithPath: filePath)
            let asset = AVAsset(url: fileURL)
            let artworkMetadata = AVMetadataItem.metadataItems(
                from: asset.commonMetadata,
                filteredByIdentifier: .commonIdentifierArtwork
            )

            guard let artworkItem = artworkMetadata.first else {
                DispatchQueue.main.async { result(nil) }
                return
            }

            if let imageData = artworkItem.dataValue {
                DispatchQueue.main.async { result(FlutterStandardTypedData(bytes: imageData)) }
            } else if let image = artworkItem.value as? UIImage, let data = image.pngData() {
                DispatchQueue.main.async { result(FlutterStandardTypedData(bytes: data)) }
            } else if let image = artworkItem.value as? UIImage, let data = image.jpegData(compressionQuality: 1.0) {
                DispatchQueue.main.async { result(FlutterStandardTypedData(bytes: data)) }
            } else if let value = artworkItem.value as? Data {
                DispatchQueue.main.async { result(FlutterStandardTypedData(bytes: value)) }
            } else {
                DispatchQueue.main.async { result(nil) }
            }
        }
    }

    private func getAlbumArtMimeType(filePath: String, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInitiated).async {
            let fileManager = FileManager.default
            guard fileManager.fileExists(atPath: filePath) else {
                DispatchQueue.main.async { result(nil) }
                return
            }

            let fileURL = URL(fileURLWithPath: filePath)
            let asset = AVAsset(url: fileURL)
            let artworkMetadata = AVMetadataItem.metadataItems(
                from: asset.commonMetadata,
                filteredByIdentifier: .commonIdentifierArtwork
            )

            guard let artworkItem = artworkMetadata.first,
                  let imageData = artworkItem.dataValue else {
                DispatchQueue.main.async { result(nil) }
                return
            }

            let mimeType = self.detectMimeType(from: imageData)
            DispatchQueue.main.async { result(mimeType) }
        }
    }

    private func detectMimeType(from data: Data) -> String? {
        guard data.count >= 4 else { return nil }
        var header = [UInt8](repeating: 0, count: 4)
        data.copyBytes(to: &header, count: 4)

        if header[0] == 0xFF && header[1] == 0xD8 && header[2] == 0xFF { return "image/jpeg" }
        if header[0] == 0x89 && header[1] == 0x50 && header[2] == 0x4E && header[3] == 0x47 { return "image/png" }
        if header[0] == 0x47 && header[1] == 0x49 && header[2] == 0x46 && header[3] == 0x38 { return "image/gif" }
        if header[0] == 0x42 && header[1] == 0x4D { return "image/bmp" }
        if data.count >= 12 {
            var webp = [UInt8](repeating: 0, count: 12)
            data.copyBytes(to: &webp, count: 12)
            if webp[0] == 0x52 && webp[1] == 0x49 && webp[2] == 0x46 && webp[3] == 0x46 &&
               webp[8] == 0x57 && webp[9] == 0x45 && webp[10] == 0x42 && webp[11] == 0x50 {
                return "image/webp"
            }
        }
        if (header[0] == 0x49 && header[1] == 0x49 && header[2] == 0x2A && header[3] == 0x00) ||
           (header[0] == 0x4D && header[1] == 0x4D && header[2] == 0x00 && header[3] == 0x2A) {
            return "image/tiff"
        }
        return "image/jpeg"
    }
}

// MARK: - AVMetadataItem data extraction (artwork as Data)
extension AVMetadataItem {
    var dataValue: Data? {
        guard let value = value else { return nil }
        if let data = value as? Data { return data }
        if let string = value as? String, let data = Data(base64Encoded: string) { return data }
        if let dict = value as? [String: Any], let data = dict["data"] as? Data { return data }
        return nil
    }
}

File: android/src/main/kotlin/com/music_feature_analyzer/MusicFeatureAnalyzerPlugin.kt
package com.music_feature_analyzer

import android.content.Context
import android.media.MediaMetadataRetriever
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File
import java.util.concurrent.Executors

/** MusicFeatureAnalyzerPlugin */
class MusicFeatureAnalyzerPlugin : FlutterPlugin, MethodCallHandler {
    private val TAG = "MusicFeatureAnalyzerPlugin"
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val handler = Handler(Looper.getMainLooper())
    private val executor = Executors.newCachedThreadPool()

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        Log.i(TAG, "onAttachedToEngine: Attaching MusicFeatureAnalyzerPlugin")
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.music_feature_analyzer/audio_metadata")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
        Log.i(TAG, "onAttachedToEngine: Channel initialized")
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        if (call.method == "verifyConnection") {
            // Log that we received the verification call
            Log.i(TAG, "verifyConnection: Received verification request")
            
            val loadingMode = detectLoadingMode()
            val response = mapOf(
                "connected" to true,
                "channelName" to "com.music_feature_analyzer/audio_metadata",
                "loadingMode" to loadingMode.name,
                "handlerSet" to true,
                "platform" to "Android",
                "pluginVersion" to "1.0.1-beta-07"
            )
            result.success(response)
            Log.i(TAG, "verifyConnection: Responded success")
            return
        }

        val path = call.argument<String>("path")
        if (path == null) {
            result.error("INVALID_ARGUMENT", "File path is required", null)
            return
        }

        when (call.method) {
            "getMetadata" -> executor.execute { getMetadata(path, result) }
            "getAlbumArt" -> executor.execute { getAlbumArt(path, result) }
            "getAlbumArtMimeType" -> executor.execute { getAlbumArtMimeType(path, result) }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        Log.i(TAG, "onDetachedFromEngine: Detaching MusicFeatureAnalyzerPlugin")
        channel.setMethodCallHandler(null)
    }

    private fun getMetadata(filePath: String, result: Result) {
        val retriever = MediaMetadataRetriever()
        val metadata = mutableMapOf<String, Any?>()

        try {
            val file = File(filePath)
            if (file.exists()) {
                val fileSizeLong = file.length()
                metadata["fileSize"] = if (fileSizeLong > Int.MAX_VALUE) Int.MAX_VALUE else fileSizeLong.toInt()
            } else {
                metadata["error"] = "File does not exist: $filePath"
                handler.post { result.success(metadata) }
                return
            }

            var dataSourceSet = false
            try {
                retriever.setDataSource(filePath)
                dataSourceSet = true
            } catch (e: IllegalArgumentException) {
                try {
                    retriever.setDataSource(file.absolutePath)
                    dataSourceSet = true
                } catch (e2: Exception) {
                    try {
                        retriever.release()
                    } catch (ignore: Exception) {}
                    metadata["error"] = "IllegalArgumentException in setDataSource: ${e2.message}"
                    metadata["mimeType"] = getMimeTypeFromExtension(filePath)
                    handler.post { result.success(metadata) }
                    return
                }
            } catch (e: RuntimeException) {
                try {
                    retriever.release()
                } catch (ignore: Exception) {}
                metadata["error"] = "RuntimeException in setDataSource: ${e.message}"
                metadata["mimeType"] = getMimeTypeFromExtension(filePath)
                handler.post { result.success(metadata) }
                return
            }

            if (!dataSourceSet) {
                try {
                    retriever.release()
                } catch (ignore: Exception) {}
                metadata["error"] = "Failed to set data source"
                metadata["mimeType"] = getMimeTypeFromExtension(filePath)
                handler.post { result.success(metadata) }
                return
            }

            try {
                val title = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_TITLE)
                metadata["title"] = if (title == null || title == "null") null else title
                
                val artist = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ARTIST)
                metadata["artist"] = if (artist == null || artist == "null") null else artist
                
                val album = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ALBUM)
                metadata["album"] = if (album == null || album == "null") null else album
                
                val albumArtist = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ALBUMARTIST)
                metadata["albumArtist"] = if (albumArtist == null || albumArtist == "null") null else albumArtist
                
                val genre = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_GENRE)
                metadata["genre"] = if (genre == null || genre == "null") null else genre
                
                val year = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_YEAR)
                metadata["year"] = if (year == null || year == "null") null else year
                val date = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DATE)
                metadata["date"] = if (date == null || date == "null") null else date
                
                val composer = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_COMPOSER)
                metadata["composer"] = if (composer == null || composer == "null") null else composer
                
                val writer = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_WRITER)
                metadata["writer"] = if (writer == null || writer == "null") null else writer
                
                val trackNumber = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_CD_TRACK_NUMBER)
                metadata["trackNumber"] = if (trackNumber == null || trackNumber == "null") null else trackNumber
                
                val discNumber = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DISC_NUMBER)
                metadata["discNumber"] = if (discNumber == null || discNumber == "null") null else discNumber

                val durationStr = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
                metadata["duration"] = durationStr?.toLongOrNull()

                // METADATA_KEY_BITRATE is in bits/sec; convert to kbps for Dart contract
                val bitrateStr = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_BITRATE)
                val bps = bitrateStr?.toLongOrNull()
                metadata["bitrate"] = if (bps != null && bps > 0) (bps / 1000).toInt() else null

                val mimeType = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_MIMETYPE)
                metadata["mimeType"] = if (mimeType != null && mimeType != "null" && mimeType.isNotEmpty()) {
                    mimeType
                } else {
                    getMimeTypeFromExtension(filePath)
                }

                val embeddedPicture = retriever.embeddedPicture
                metadata["hasAlbumArt"] = embeddedPicture != null && embeddedPicture.isNotEmpty()
            } catch (e: Exception) {
                metadata["error"] = "Exception during metadata extraction: ${e.message}"
                metadata["mimeType"] = getMimeTypeFromExtension(filePath)
            }

            try {
                retriever.release()
            } catch (ignore: Exception) {}
            handler.post { result.success(metadata) }
        } catch (e: Exception) {
            try {
                retriever.release()
            } catch (ex: Exception) {
            }
            metadata["error"] = "Exception in getMetadata: ${e.javaClass.simpleName}: ${e.message}"
            if (!metadata.containsKey("mimeType")) {
                metadata["mimeType"] = getMimeTypeFromExtension(filePath)
            }
            handler.post { result.success(metadata) }
        }
    }

    private fun getMimeTypeFromExtension(filePath: String): String {
        val extension = filePath.substringAfterLast('.', "").lowercase()
        return when (extension) {
            "mp3" -> "audio/mpeg"
            "m4a", "m4p" -> "audio/mp4"
            "aac" -> "audio/aac"
            "wav" -> "audio/wav"
            "flac" -> "audio/flac"
            "ogg" -> "audio/ogg"
            "wma" -> "audio/x-ms-wma"
            "opus" -> "audio/opus"
            "aiff", "aif" -> "audio/aiff"
            "alac" -> "audio/alac"
            "m4b" -> "audio/mp4"
            "amr" -> "audio/amr"
            "3ga" -> "audio/3gpp"
            else -> "audio/mpeg" // Default fallback
        }
    }

    private fun getAlbumArt(filePath: String, result: Result) {
        val retriever = MediaMetadataRetriever()
        try {
            val file = File(filePath)
            if (!file.exists()) {
                try {
                    retriever.release()
                } catch (ignore: Exception) {}
                handler.post { result.success(null) }
                return
            }

            try {
                retriever.setDataSource(filePath)
            } catch (e: Exception) {
                try {
                    retriever.release()
                } catch (ignore: Exception) {}
                handler.post { result.success(null) }
                return
            }
            
            val art = retriever.embeddedPicture
            if (art == null || art.isEmpty()) {
                try {
                    retriever.release()
                } catch (ignore: Exception) {}
                handler.post { result.success(null) }
                return
            }

            try {
                retriever.release()
            } catch (ignore: Exception) {}
            handler.post { result.success(art) }
        } catch (e: Exception) {
            try {
                retriever.release()
            } catch (ex: Exception) {
            }
            handler.post { result.success(null) }
        }
    }

    private fun getAlbumArtMimeType(filePath: String, result: Result) {
        val retriever = MediaMetadataRetriever()
        try {
            val file = File(filePath)
            if (!file.exists()) {
                try {
                    retriever.release()
                } catch (ignore: Exception) {}
                handler.post { result.success(null) }
                return
            }
            
            try {
                retriever.setDataSource(filePath)
            } catch (e: Exception) {
                try {
                    retriever.release()
                } catch (ignore: Exception) {}
                handler.post { result.success(null) }
                return
            }
            
            val art = retriever.embeddedPicture
            try {
                retriever.release()
            } catch (ignore: Exception) {}

            if (art != null && art.size >= 4) {
                val mimeType = when {
                    art[0] == 0xFF.toByte() && art[1] == 0xD8.toByte() && art[2] == 0xFF.toByte() -> "image/jpeg"
                    art[0] == 0x89.toByte() && art[1] == 0x50.toByte() &&
                            art[2] == 0x4E.toByte() && art[3] == 0x47.toByte() -> "image/png"
                    art[0] == 0x47.toByte() && art[1] == 0x49.toByte() &&
                            art[2] == 0x46.toByte() && art[3] == 0x38.toByte() -> "image/gif"
                    art.size >= 12 && art[0] == 0x52.toByte() && art[1] == 0x49.toByte() &&
                            art[2] == 0x46.toByte() && art[3] == 0x46.toByte() &&
                            art[8] == 0x57.toByte() && art[9] == 0x45.toByte() &&
                            art[10] == 0x42.toByte() && art[11] == 0x50.toByte() -> "image/webp"
                    art[0] == 0x42.toByte() && art[1] == 0x4D.toByte() -> "image/bmp"
                    else -> "image/jpeg"
                }
                handler.post { result.success(mimeType) }
            } else {
                handler.post { result.success(null) }
            }
        } catch (e: Exception) {
            try {
                retriever.release()
            } catch (ex: Exception) {
            }
            handler.post { result.success(null) }
        }
    }
    
    // Enum for loading detection
    enum class LoadingMode {
        PUBLISHED_PACKAGE,
        LOCAL_PATH,
        UNKNOWN
    }
    
    private fun detectLoadingMode(): LoadingMode {
        // Simple placeholder as reliable detection is complex and not strictly necessary for functionality
        return LoadingMode.UNKNOWN
    }
}
